Forager with diode-free Partial Heawood wiring ("Acid")

peterjc opened the pull request:
Yes, yet _another_ alternative PCB draft pull request. I'm hoping for comments/feedback (even if to say you its too different so you won't merge it and I should give it a different name). Or, you are of course welcome to merge (or squash and merge) now or later -  I'll turn it into a non-draft pull request once I've built it.

This is based on my work in #19 designing a diode-free direct wiring reversible PCB using the Xiao NRF52840 _plus_ which has 19 usable GPIO pins. That is nice and simple, but the _tiny_ castellation soldering will be tricky. It will be quicker with no diodes, but argubly harder? I'm not sure I'm ready to try that without help 😉 

This PR is an alternative diode-free _graph-theory_ reversible PCB using the _original_ Xiao NRF52840, and the extra pair of NFC pins on the back making 11+2=13 GPIOs. See ZMK setting `CONFIG_NFCT_PINS_AS_GPIOS=y` for using these. That means it probably won't work with Xiao clones, as they often have alternative extra pads on the back.

<img width="1457" height="1231" alt="Screenshot 2025-10-10 at 09 10 01" src="https://github.com/user-attachments/assets/701261a1-1413-4ca5-9855-3c50d9e87a1a" />

Compared to the original Forager PCB, the castellation soldering is the same, but now rather than just one rear pad to solder there are three (battery and the two NFC pads). That's a little more work with through board soldering, but with no diodes I think this will be much easier and quicker overall. The cornermost NFC pad P0.10 is very awkwardly placed (almost below D7/D14). The almost triangular cutout I used is to allow a trace to reach this pad without crossing the radio keepout.

<img width="1487" height="977" alt="Screenshot 2025-10-10 at 09 23 19" src="https://github.com/user-attachments/assets/60e11b6e-d22d-4697-8cdd-51943e5fd680" />
<img width="1515" height="1181" alt="Screenshot 2025-10-10 at 09 56 59" src="https://github.com/user-attachments/assets/6d8054d5-1c2a-4fed-a25f-bc6228c6d2d0" />
<img width="1516" height="1173" alt="Screenshot 2025-10-10 at 09 57 19" src="https://github.com/user-attachments/assets/63301527-72f9-4b42-8f7c-713dc308b956" />

The wiring is essentially a cut down version of https://github.com/triliu/Heawood42 which uses the full Heawood Graph (14 nodes in a 7x7 sparse scanning matrix for 21 edges or keys on each half keyboard). I wrote about using the [Heawood Graph for Keyboard wiring](https://astrobeano.blogspot.com/2025/05/topology-meets-custom-keyboard-circuit.html) in a blog post. Here it is a Partial Heawood Graph (13 nodes in a 6x7 sparse scanning matrix for 18 edges or keys - one spare).

Space is less constrained than with direct wiring, so to try to keep the firmware and wiring simple, I could link like pins on both sides. For example Qwerty `Q` and `P` are both on scanning column zero (net `CO`, NFC pad `P0.10`) and scanning row zero (net `R0`, pad `D5`).

The weakness of this graph wiring is it is only 4-key roll-over per half (you can do better with more GPIOs per switch, eg the [Tutte Coxeter Graph for keyboard wiring](https://astrobeano.blogspot.com/2025/07/my-first-keyboard-pcb-design-diode-free.html)).

The name "Acid" is from Partial Heawood v1.0.0 --> pH 1.0 --> Acid.

This PR includes the power switch (#14 ), large battery cutout (#18), and Molex PicoBlade battery connector as on #19.

I have started ordering parts including a pair of the _original_ Xiao controllers, batteries, and battery connectors to confirm how they fit together and help me verify the planned placements.

peterjc committed:
Save forager-pcb.kicad_pcb as forager-ph.kicad_pcb in KiCad v9.0.2

This will be the Forager Partial Heawood or "Acid" PCB, still using
the Seeed XIAO nRF52840 Plus controller but also the two NFC pads
for 11+2 GPIOs for a diode-free reversible Graph Theory wiring.

peterjc committed:
Remove all the old switch traces, set nets for direct wiring

Set all the right hand switch diode nets to RGND

Used sed to search and replace, and then opened and resaved in
KiCad which renumbered the now much smaller list of nets:

sed -i.bak 's/net .. "Net-(RD.*-A)/net 33 "RGND/g' forager-plus-pcb.kicad_pcb

Set all left hand switch diode nets to LGND. Again, used sed
and then opened & saved in KiCad to renumber:

sed -i.bak 's/net . "Net-(LD.*-A)/net 1 "LGND/g' forager-plus-pcb.kicad_pcb
sed -i.bak 's/net .. "Net-(LD.*-A)/net 1 "LGND/g' forager-plus-pcb.kicad_pcb

Swap switch column nets to zmk-helpers style, did rest manually:

sed -i.bak -e 's#"/RCOL0"#"RM4"#g' -e 's#"/RCOL1"#"RM3"#g' \
    -e 's#"/RCOL2"#"RM2"#g' -e 's#"/RCOL3"#"RM1"#g' \
    -e 's#"/RCOL4"#"RM0"#g' forager-plus-pcb.kicad_pcb

sed -i.bak -e 's#"/LCOL0"#"LM4"#g' -e 's#"/LCOL1"#"LM3"#g' \
    -e 's#"/LCOL2"#"LM2"#g' -e 's#"/LCOL3"#"LM1"#g' \
    -e 's#"/LCOL4"#"LM0"#g' forager-plus-pcb.kicad_pcb

Done manually in KiCad v9.0.2

peterjc committed:
Make into a reversible board (replace switches in right, delete left)

Double sided using original Xiao BLE (11+2 GPIOs), wiring to follow.

Using ceoloide's ergogen-footprints/switch_choc_v1_v2.js footprint
in v1 only mode (not enough space near controller for v2).

Use 2-pin horizontal SMD PicoBlade MX 1.25mm battery connector.
The inaccurate term JST 1.25mm, or 'Micro JST 1.25mm' is often
used for these. The connector socket is under 4mm so should fit,
and this is commonly offered as the connector with LiPo batteries.

Battery cutout by changing the PCB outline

Move battery connector forward to avoid bottom case rib

peterjc committed:
Use XIAO-nRF52840 v15.step

See https://files.seeedstudio.com/wiki/XIAO-BLE/seeed-studio-xiao-nrf52840-3d-model.zip
from https://wiki.seeedstudio.com/XIAO_BLE/#seeed-studio-xiao-nrf52840

peterjc committed:
Add P0.09 and P0.10 NFC pads to footprint

These are in the wireless keepout (sigh), and P0.10 is so tight in the
corner it will be tricky.

peterjc committed:
Partial Heawood Graph Theory diode-free switch wiring done

peterjc committed:
Lengthen battery cutout

peterjc committed:
Rotate battery connection and move back to give clearance,

Beware the bottom case ribs...

peterjc committed:
Show 15x35mm battery placeholder

peterjc committed:
Restore direct solder points for battery connection

peterjc committed:
Add bonus sideways SMD thumb switch to fill out the matrix.

This will need a tiny case mode, and perhaps a longer than usual button
if this isn't close enough to the edge? Vertical position limited by
bottom case rib.

Wiring required a few more vias for the ground plane.

peterjc committed:
Change to thru-hole battery terminals for direct solder option

Hat tip https://github.com/letmegobacktosleep/forager/tree/power-switch/forager-pcb

peterjc committed:
Add MSK12C02 power switch - LCSC C431540 or similar, SMD

These should both be up off (towards controller corner), and
down on (away from controller). Ought to be able to solder
bridge the pads if not using the switch?

Hat tip https://github.com/letmegobacktosleep/forager/tree/power-switch/forager-pcb

peterjc committed:
Enlarge most 2x2.6mm hotswap pads to 2.6x2.6mm

This is more symetric and more closely matches the original Forager
hotswap footprint. Unclear why the Ceoloide Ergogen hotswaps are
that small on one side - although they work fined fine for me as is
on another board.

A few are still on the small side due to constaints like the battery
cut out, or wiring.

peterjc committed:
Graph Theory Silkscreen art using Heawood Graph

peterjc committed:
Set title and revision: Forager Acid 1.0.0

peterjc committed:
Gerbers for Forager Acid 1.0.0

peterjc commented:
I [blogged](https://www.blogger.com/blog/post/edit/7275378574543326675/3066355662379924906) about this - the PCBs and case have been shipped, and I think I have all the parts now 🤞

MSmaili commented:
> I [blogged](https://www.blogger.com/blog/post/edit/7275378574543326675/3066355662379924906) about this - the PCBs and case have been shipped, and I think I have all the parts now 🤞

The link you shared does not work!

peterjc committed:
Re-inforce BAT+ connection with trace on both sides.

This may help if the copper is damaged while soldering, or when the
board is drilled during fabrication.

peterjc commented:
Mixed news, mostly my inexperience with soldering to blame.

My left "Acid" board with only with the controller soldered on) seems to work but after a few minutes drops off the USB and stops responding. Some sort of power issue is my guess, and macOS cuts it off? Probably bad soldering somewhere. I may need a spare controller...

My right board was looking good (battery not tested yet) aside something odd with the middle index keys that feels like a typo in my firmware matrix. However, after I reflowed a couple of not so pretty pads, soldered on the nuts, and assembled it all, some keys had stopped working (flashed it as the primary right half for solo testing). Time to take it apart again...

![Top side of right half of Forager Acid keyboard without switches, blue PCB visible through switch holes in white top case](https://github.com/user-attachments/assets/f89713a4-54bd-48ee-96d7-a48db8415795)

![Underside of right half of Forager Acid keyboard with blue PCB visible through clear bottom case](https://github.com/user-attachments/assets/ae2be1b7-0d7b-4697-ad68-9d49ecc0e50a)

The silkscreen art could be better placed for a transparent bottom case (especially the little Heawood Graph, name and version). They probably would look fine on the top which was my priority, but I order the top case printed in white in error.

The power switch seems [not to stick out enough](https://github.com/carrefinho/forager/issues/14#issuecomment-3510873081).

The bonus thumb key using a reset button isn't going to be practical unless perhaps pushed via the case itself (shown placed but not soldered in this photo). I think I'll drop that on any revision to the PCB:

![Sidefacing reset button does not stick out enough to use it](https://github.com/user-attachments/assets/c2d8baeb-b456-4c4c-ac55-61f2a2106b6d)

The outer diameter of my nuts seems to be a fraction small - workable, but harder to get them perfectly vertical and centered. They still worked to assemble it.

Both the left and right bottom cases show a slight warped in the middle of the flat panel above the battery.

P.S. Getting the metal dowels into the tenting legs in a pain - just too tight on this print. And my superglue for the magnets inside the case had dried up, so I have not tried them properly yet.

peterjc commented:
Good news: Spare controller arrived, and the left half now also done. I am using the "Forager Acid" to type this now, with MagSafe tenting (stick on metal rings and UGreen stands) and [3D printed Chicago Stenographer keycaps](https://github.com/infused-kim/kb_keycaps_chicago_stenographer)!

![2025-11-26 11 19 20](https://github.com/user-attachments/assets/73f436f3-c504-4915-85d5-53c248424e50)

The surface mounted battery connector and power switch were much easier to solder this time (second attempt), and I am comfortable doing hotswaps now (having done about 100 in total at this point).

Mixed news: I still found surface mounting the main Xioa pins rather tricky (repeat rounds of soldering, circuit testing, and resoldering). I actually found the through cut-out soldering for the NFC pins and B+ easier, so this is still probably easier that the original Forager PCB with the diodes.

Bad news: The power switch is on straight this time, but still doesn't stick out enough to use without a tool. I can only assume that some versions of this switch have a longer handle.

Also the reset switch on my right half seems to have stopped working, so I am dependent on ZMK working and putting `&bootloader` on the right half via ZMK Studio when I want to reflash (eg change my combos, although one day ZMK Studio should support that).

--

@carrefinho What do you think? Is this something you'd consider merging?

If yes, what changes would you request or suggest? I am thinking to remove the 5V and 3v3 pads to reduce the chances of getting a short-circuit through amateur soldering (like on my failed PCB), and flip the battery connection polarity (to match my EHAO branded LP501535 batteries). Where should I add any Acid specific build guide?

If no, I will make the Acid branch the default on my fork of the repository and add some notes to the main README about the differences.

--

Edit: Typos

peterjc commented:
Looking at https://docs.beekeeb.com/toucan-keyboard/battery-specs the Forager Acid v1.0.0 PCB seems to use the same orientation as the Toucan (Molex PicoBlade 51021-0200, 1.25 mm pitch), and my battery matched their "No Go" photo:

![battery connections for Toucan](https://docs.beekeeb.com/~gitbook/image?url=https%3A%2F%2F310059516-files.gitbook.io%2F%7E%2Ffiles%2Fv0%2Fb%2Fgitbook-x-prod.appspot.com%2Fo%2Fspaces%252F-MipCDWiRJbsrF2klOWK%252Fuploads%252FGBarFpihQwsb98R3nV2x%252Ftoucan-card.png%3Falt%3Dmedia%26token%3Dc86639ce-c9fb-48ae-a5c6-6dd3cb99397c&width=768&dpr=1&quality=100&sign=1fc72ee4&sv=2)

Their guide also touches on the possibility of [changing battery connection polarity](https://docs.beekeeb.com/toucan-keyboard/battery-specs#not-recommended-changing-the-polarity) which they do not recommend.

Perhaps putting _two_ Molex PicoBlade footprints on the PCB to cover both polarities isn't a completely crazy idea?

peterjc committed:
Bump Acid to v1.1.0, drop the unused 'bonus' thumb button

peterjc committed:
Remove unused 5V and 3V3 pads (slightly less chance to short circuit with bad soldering)

peterjc committed:
Fix edge cut under USB port (was cutting twice with two footprints - cancelled out?)

This should give access to the RESET pad for emergency use (eg reset
button broken off).

peterjc committed:
Duplicate battery connectors with alternative polarity.

Also return to original's pads for the fall back battery soldering option.
The through holes seemed very small to me.

And update the version on the silksceen.

peterjc committed:
Forager Acid gerbers v1.0.0 --> v1.1.0

peterjc commented:
Updated to v1.1.0 - this fixes the cutout under the USB for access to the reset pad (wanted that today), duplicates the battery connectors with the other polarity (see above), reverts to pads for direct battery soldering (probably easier), and drops the bonus thumb key (wasn't practical).

<img width="1330" height="1126" alt="Screenshot 2026-01-30 at 15 59 22" src="https://github.com/user-attachments/assets/525fdb26-2f79-48f2-85ed-340e91948e0c" />
<img width="1335" height="1126" alt="Screenshot 2026-01-30 at 15 59 34" src="https://github.com/user-attachments/assets/1ba2b2ff-d824-4321-9a79-cc7e8cbd3a91" />

peterjc commented:
Correction: The cutout for the reset pad etc under the USB _did_ work in v1.0.0, but didn't render right in KiCad (two cuts on top of each other). Renders right in v1.1.0.
